### ISSUE TITLE
MacOS support

### ISSUE THREAD

**Issue by cconcolato** [AOMediaCodec/libiamf#61]
I tried compiling on MacOS and couldn't get it to work, some error about `io.h` not being found.
Is MacOS supported? 
I suggest adding supported platforms explicitly to the README.

**Comment by kr728-kim** [comment]
We are working to ensure that the IAMF decoder supports MacOS.
The IAMF decoder has dependent libraries that need to be checked and are being checked.

**Comment by yilun-zhangs** [comment]
> Does this build on MacOS now? We still should update the README.
Yes, in #65 , it has fixed MacOS compile error, and I have updated README in #75

**Comment by jwcullen** [comment]
Re-opening this because we should make sure it builds with standard features. At least it should build with both binaural libraries and any supported codecs.

**Comment by trevorknight** [comment]
I have been trying to build with codecs and renderers on MacOs and wanted to just share some tips and troubles.

## Building with codecs but without renderers

I was able to build without binaural renders but with AAC/Flac/Opus codecs with the following technique

1. Install homebrew
2. Install `$ brew install cmake opus fdk-aac flac`
3. Clone libiamf
4. Delete the contents of libiamf/code/dep_codecs/lib/
  ```
  cd libiamf/code/dep_codecs/lib/
  rm *
  cd ..
  ```
6. Copy the libs
  ```
  $ cp /opt/homebrew/opt/flac/lib/* ./dep_codecs/lib/
  $ cp /opt/homebrew/opt/fdk-aac/lib/* ./dep_codecs/lib/
  $ cp /opt/homebrew/opt/opus/lib/* ./dep_codecs/lib/
  ```
7. build
  ```
  $ cmake -DCMAKE_INSTALL_PREFIX=${PWD}/build_libs  -DMULTICHANNEL_BINAURALIZER=OFF -DHOA_BINAURALIZER=OFF .
  $ make
  $ make install
  ```
8. Make iamfdec
  ```
  $ cd test/tools/iamfdec
  $ cmake -DCMAKE_INSTALL_PREFIX=${PWD}/../../../build_libs  -DMULTICHANNEL_BINAURALIZER=OFF -DHOA_BINAURALIZER=OFF .
  $ make

  ```  
  

## Issues trying to build with renderers

For code/dep_external/src/binaural/build.sh:
- Maybe want to `set -e` so the script exits on error
- VISR is now in [a branch of the ebu/bear repo](https://github.com/ebu/bear/tree/visr), which has been patched since that release, which helps some things. Can still put it in that folder if you want with `git clone -b visr --single-branch https://github.com/ebu/bear.git rd-audio-visr-public-master`
- I wasn't able to successfully use/build boost using the technique of the script, I ran into errors building.
- I was able to build BEAR/VISR/EAR etc from a `brew install boost`, so could just encourage that then use CMake's `find_package(Boost`
- If `-DBUILD_PYTHON_BINDINGS=ON` was changed to `-DBUILD_PYTHON_BINDINGS=OFF` for both VISR and BEAR builds then things are much simpler (no python dependencies) and the pybind section could be removed.
- I killed the script when something needed sudo.


Ignoring code/dep_external/src/binaural/build.sh, I was able to build Resonance and Bear with dependencies with this technique:

1. `$ brew install cmake opus fdk-aac flac boost`
2. Create a folders for builds and for repo clones
  ```
  $ cd ~
  $ mkdir -p libiamf_deps/bear
  $ export CMAKE_INSTALL_PREFIX=$PWD/libiamf_deps/bear/
  $ export CMAKE_PREFIX_PATH=$PWD/libiamf_deps/bear/
  $ mkdir github_clones && cd github_clones
  ```
3.  Starting with VISR
  ```
  $ git clone -b visr --single-branch https://github.com/ebu/bear.git visr
  $ cd visr
  $ cmake -B build . -DBUILD_PYTHON_BINDINGS=OFF -DBUILD_DOCUMENTATION=OFF -DBUILD_AUDIOINTERFACES_PORTAUDIO=OFF -DBUILD_USE_SNDFILE_LIBRARY=OFF
  $ cmake --build build -t install
  ```

4. Cloning BEAR
  ```
  $ cd ~/github_clones
  $ git clone https://github.com/ebu/bear.git
  $ cd bear
  $ git submodule update --init --recursive
  $ cd visr_bear
  ```

5.  In visr_bear/CMakeLists.txt, I had to remove this section as there were build errors in the unit tests
  ```
  if(BEAR_UNIT_TESTS)
    include(CTest)
    add_subdirectory(test)
  endif()
  ```

6. Build BEAR
  ```
  cmake -B build . -DBUILD_PYTHON_BINDINGS=OFF
  cmake --build build -t install
  ```

7. Resonance 
  ```
  cd ~
  mkdir -p libiamf_deps/resonance/
  export CMAKE_INSTALL_PREFIX=$PWD/libiamf_deps/resonance/
  export CMAKE_PREFIX_PATH=$PWD/libiamf_deps/resonance/
  cd ~/github_clones
  git clone https://github.com/resonance-audio/resonance-audio
  cd resonance-audio
  ./third_party/clone_core_deps.sh
  ```
  
 8. I then had to update the C++ version for Resonance because of the newer version of xsimd that uses C++14 features.  So in CMakeLists.txt change
   `set(CMAKE_CXX_STANDARD 11)`
   to
   `set(CMAKE_CXX_STANDARD 17)` (14 probably works too)
   
  9. Build Resonance
    ```
    $ ./build.sh -t=RESONANCE_AUDIO_API -p=Release
    ```
  

And then I got stuck building iamf2bear.  

```
$ cd ~/github_clones
$ git clone https://github.com/AOMediaCodec/libiamf
$ cd1 libiamf/code
$ cd dep_codecs/lib/
$ rm *
$ cd ../..
$ cd ~
$ export CMAKE_INSTALL_PREFIX=$PWD/libiamf_deps/bear/
$ cd github_clones/libiamf/code/dep_external/src/binaural/iamf2bear/
$ cmake .
$ make
```

For some reason make was not resolving the includes properly despite these lines in code/dep_external/src/binaural/iamf2bear/CMakeLists.txt

```
set(CODEC_INCLUDE_DIR  "${CMAKE_INSTALL_PREFIX}/include")


include_directories(
    ${CODEC_INCLUDE_DIR}
)
```

And I ran out of enthusiasm.  

I hope this helps someone else trying to build or helps make changes to the build.sh.

**Comment by yilun-zhangs** [comment]
@trevorknight 
Thanks for your great efforts to compile externals, 
actually, we have made some changes to build.sh at local side, to let it work more efficiently.
Now it it works under linux with only one step.
We will try macOS as well.

**Comment by trevorknight** [comment]
Okay, great!  Looking forward to seeing it

**Comment by yilun-zhangs** [comment]
@trevorknight 
Thanks for your valuable suggestions again, it help me a lot prepare the PR.
Bear depends on too many other libraries, I need to make some modification.
I have submitted #110 to make the binaural compile more efficient, I will share the steps.

1. Compile dep_codecs.
$ cd dep_codecs
$ ./build.sh

2. Compile dep_external(binaural libraries)
`Please install boost firstly, please refer to [README](dep_external/src/binaural/README.md)`
`Because of network issue, default.tf downloading maybe failed, but it dose not affect the building, so pleaes remove 'set -e' in build.sh`
$ cd dep_external/src/binaural/
$ ./build.sh

3. Compile libiamf
$ cd code
$ ./build.sh

I hope this change could help you, if are there any issue, please feel free to let me know.
I'd always like to hear your opinions.

Thanks

**Comment by yilun-zhangs** [comment]
The previous PR #110 is merged and covered for some reason, I create another #112

**Comment by trevorknight** [comment]
Okay, I'll give it another try and let you know! Thank you very much

**Comment by fire** [comment]
Is mac supported now?

**Comment by olilarkin** [comment]
i've been trying to compile it this morning... hours of my life i won't get back!

**Comment by olilarkin** [comment]
I was finally able to compile it with resonance and libear on apple silicon. I had to patch libVISR, install an old version of boost.

I now have a working iamfdec cli tool with a million not very portable dylib dependencies.

**Comment by fire** [comment]
Is this something that can be a pull request to this repository?

**Comment by mormegil6** [comment]
@yilun-zhangs could you be more specific on how to install boost v1.82 correctly? I believe I successfully compiled it, but the build.sh in dep_external/src/binaural still sees the one I have installed by homebrew, v1.87. I can't figure out how to "show" the v1.82 to the build.sh script.
Also, the link to the v1.82 zip archive seems to be no longer working. I have found the source code on the boost.org website: https://www.boost.org/users/history/version_1_82_0.html

**Comment by PaulinRoman** [comment]
Hi, I have tried compiling the lib and the renderers last week but got stuck on the same points as trevorknight, and I can't figure out the changes needed on boost's version, are there any updates to the project and the build ?

**Comment by yilun-zhangs** [comment]
It has supported already,
Each commit will execute MacOS compile `action`
